godot-core/src/builtin/basis.rs

```diff
@@ -0,0 +1,980 @@
+        // compares it.
+        //
+        // In this way we can assert that both functions: basis to euler / euler to basis
+        // are correct.
+
+        // Euler to rotation
+        let original_euler: Vector3 = deg_to_rad(deg_original_euler);
+        let to_rotation: Basis = Basis::from_euler(
+            rot_order,
+            original_euler.x,
+            original_euler.y,
+            original_euler.z,
+        );
+
+        // Euler from rotation
+        let euler_from_rotation: Vector3 = to_rotation.euler_angles(rot_order);
+        let rotation_from_computed_euler: Basis = Basis::from_euler(
+            rot_order,
+            euler_from_rotation.x,
+            euler_from_rotation.y,
+            euler_from_rotation.z,
+        );
+
+        let res: Basis = to_rotation.inverse() * rotation_from_computed_euler;
+
+        assert!(
+            (res.col_x() - Vector3::new(1.0, 0.0, 0.0)).length() <= 0.1,
+            "Fail due to X {} with {deg_original_euler} using {rot_order:?}",
+            res.col_x()
+        );
+        assert!(
+            (res.col_y() - Vector3::new(0.0, 1.0, 0.0)).length() <= 0.1,
+            "Fail due to Y {} with {deg_original_euler} using {rot_order:?}",
+            res.col_y()
+        );
+        assert!(
+            (res.col_z() - Vector3::new(0.0, 0.0, 1.0)).length() <= 0.1,
+            "Fail due to Z {} with {deg_original_euler} using {rot_order:?}",
+            res.col_z()
+        );
+
+        // Double check `to_rotation` decomposing with XYZ rotation order.
+        let euler_xyz_from_rotation: Vector3 = to_rotation.euler_angles(EulerOrder::XYZ);
+        let rotation_from_xyz_computed_euler: Basis = Basis::from_euler(
+            EulerOrder::XYZ,
+            euler_xyz_from_rotation.x,
+            euler_xyz_from_rotation.y,
+            euler_xyz_from_rotation.z,
+        );
+
+        let res = to_rotation.inverse() * rotation_from_xyz_computed_euler;
+
+        assert!(
+            (res.col_x() - Vector3::new(1.0, 0.0, 0.0)).length() <= 0.1,
+            "Double check with XYZ rot order failed, due to X {} with {deg_original_euler} using {rot_order:?}",
+            res.col_x(),
+        );
+        assert!(
+            (res.col_y() - Vector3::new(0.0, 1.0, 0.0)).length() <= 0.1,
+            "Double check with XYZ rot order failed, due to Y {} with {deg_original_euler} using {rot_order:?}",
+            res.col_y(),
+        );
+        assert!(
+            (res.col_z() - Vector3::new(0.0, 0.0, 1.0)).length() <= 0.1,
+            "Double check with XYZ rot order failed, due to Z {} with {deg_original_euler} using {rot_order:?}",
+            res.col_z(),
+        );
+    }
+
+    #[test]
+    fn consts_behavior_correct() {
+        let v = Vector3::new(1.0, 2.0, 3.0);
+
+        assert!(
+            (Basis::IDENTITY * v).is_equal_approx(v),
+            "got {got}, expected {v}",
+            got = Basis::IDENTITY * v
+        );
```

Comment:
Maybe use "actual" instead of "got" 🙂 

But you don't need to explicitly mention that, the default `assert` message already prints both sides. Instead, focus on the context:
```rs
let actual = ...;
let expected = ...;
assert!(actual, expected, "identity basis * vector");
```

Reply:
You're thinking of `assert_eq`, `assert` does not work this way. But this is the exact purpose for which i made `assert_eq_approx`, so i'll just use that.

Reply:
Yes, sorry. Using `assert_eq_approx` sounds good!